test/antrea/networkpolicy_controller.dl

```diff
@@ -0,0 +1,364 @@
+		.name       = appliedToGroupUID,
+		.uid        = k8s.UID{appliedToGroupUID},
+		.selector   = groupSelector
+	}
+}
+
+function toAntreaPeer(peers: Vec<k8s.NetworkPolicyPeer>, np: k8s.NetworkPolicy): (NetworkPolicyPeer, Vec<AddressGroup>) =
+{
+	// Empty NetworkPolicyPeer is supposed to match all addresses.
+	// See https://kubernetes.io/docs/concepts/services-networking/network-policies/#default-allow-all-ingress-traffic.
+	// It's treated as an IPBlock "0.0.0.0/0".
+	if (vec_is_empty(peers)) {
+		(matchAllPeer(), vec_empty())
+	} else {
+        var ipBlocks: Vec<IPBlock> = vec_empty();
+        var addressGroupNames: Vec<string> = vec_empty();
+        var addressGroups: Vec<AddressGroup> = vec_empty();
+
+        for (peer in peers) {
+            // A networking.NetworkPolicyPeer will either have an IPBlock or a
+            // podSelector and/or namespaceSelector set.
+            match (peer.ipBlock) {
+                Some{peerIPBlock} -> {
+                    match (toAntreaIPBlock(peerIPBlock)) {
+                        Left{err} -> {
+                            log_error("Failure processing NetworkPolicy ${np.namespace}/${np.name} IPBlock ${peerIPBlock}: ${err}")
+                        },
+                        Right{ipBlock} -> {
+                            vec_push(ipBlocks, ipBlock)
+                        }
+                    }
+                },
+                None -> {
+                    var addressGroup = createAddressGroup(peer, np);
+                    vec_push(addressGroups, addressGroup);
+                    vec_push(addressGroupNames, addressGroup.name)
+                }
+            }
+        };
+        var nppeer = NetworkPolicyPeer {
+            .addressGroups  = addressGroupNames,
+            .ipBlocks       = ipBlocks
+        };
+        (nppeer, addressGroups)
+    }
+}
+
+// toAntreaProtocol converts a v1.Protocol object to an Antrea Protocol object.
+function toAntreaProtocol(npProtocol: Option<k8s.Protocol>): Protocol =
+{
+	// If Protocol is unset, it must default to TCP protocol.
+    match (npProtocol) {
+        None -> protocolTCP(),
+        Some{proto} -> proto
+    }
+}
+
+// toAntreaServices converts a networkingv1.NetworkPolicyPort object to an
+// Antrea Service object.
+function toAntreaServices(npPorts: Vec<k8s.NetworkPolicyPort>): Vec<Service> =
+{
+	var antreaServices: Vec<Service> = vec_empty();
+	for (npPort in npPorts) {
+		var antreaService = Service {
+			.protocol   = toAntreaProtocol(npPort.protocol),
+		    .port       = match (npPort.port) {
+                None    -> None: Option<signed<32>>,
+                // TODO(abhiraut): Retrieve ports for named ports.
+                Some{p} -> Some{ k8s.intOrStringIntValue(p) }
+            }
+		};
+		vec_push(antreaServices, antreaService)
+	};
+	antreaServices
+}
+
+// toAntreaIPBlock converts a networkingv1.IPBlock to an Antrea IPBlock.
+function toAntreaIPBlock(ipBlock: k8s.IPBlock): Either<string,IPBlock> =
```

Comment:
How about a type for error too?

Reply:
In fact I would create a type like the Rust Result.

Reply:
yes, that's the way to go; I'll do that.